### PR TITLE
Rewrite

### PR DESCRIPTION
**Pull request by nikku**
### Proposed Changes

This PR aims to improve the following:

* significantly improve the inner workings of the editor
* properly configure language _and_ autocompletion _together_ (preventing [silly bugs](https://github.com/camunda/camunda-modeler/issues/3983))
* improving completion UX in various small ways
* fixing completion bugs
* making built-ins pluggable 
* prevent bogus snippet completions within paths and/or identifiers

capture 534KSo_optimized

You can try out the changes via 

```
npx @bpmn-io/sr bpmn-io/feel-editor#rewrite
```

Related to https://github.com/camunda/camunda-modeler/issues/3983

### Checklist

To ensure you provided everything we need to look at your PR:

* [x] **Brief textual description** of the changes present
* [x] **Visual demo** attached
* [x] **Steps to try out** present, i.e. [using the `@bpmn-io/sr` tool](https://github.com/bpmn-io/sr)
* [x] Related issue linked via `Closes {LINK_TO_ISSUE}` or `Related to {LINK_TO_ISSUE}`

### CODE REVIEW

File: test/spec/CodeEditor.spec.js [hunk 1]
```diff
@@ -692,8 +704,13 @@ return
...
 
-        expect(completions).not.to.be.empty;
-        expect(completions[0].label).to.have.eql('abs(n)');
+        expect(completions).to.have.length.gte(2);
+
+        // variable completions offered
+        expect(completions[0].label).to.have.eql('ab');
+
+        // followed by function completions
+        expect(completions[1].label).to.have.eql('abs(n)');
```
Comment (barmac) [inline test/spec/CodeEditor.spec.js:647]:
```suggestion
        // variable completions offered
        expect(completions[0].label).to.eql('ab');

        // followed by function completions
        expect(completions[1].label).to.eql('abs(n)');
```

Comment (nikku) [reply test/spec/CodeEditor.spec.js]: Thanks! Found and fixed additional instances -https://github.com/bpmn-io/feel-editor/pull/56/commits/8ae2242e408b8dadc68290d95ef1010bf2ea0b0d.

File: src/autocompletion/variable.js [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+function getBoost(option, match) {
+
+  const {
+    label,
+    boost
+  } = option;
+
+  if (!label.includes(match)) {
+    return -100;
+  }
```
[the remark this answers is not in the document]
Comment (nikku) [reply src/autocompletion/variable.js]: Restored via https://github.com/bpmn-io/feel-editor/pull/56/commits/e3477fbb38f76f10e12c6ad3c5fd9c2fd2e290ed.

**Comment by nikku** [comment]
Ready for review (again). Please have a look!

**Review by barmac — APPROVED** [review]

**Comment by barmac** [comment]
thumbsup

**Pull request merged**